src/reflection.ts

```diff
@@ -358,7 +358,7 @@ export class Parser {
       const discriminatorFieldName = `${fieldName}_type`;
       const unionDiscriminator = this.getField(discriminatorFieldName, typeIndex);
 
-      return this.readUnionLambda(field, unionDiscriminator, readDefaults);
+      return this.readUnionLambda(field, unionDiscriminator, typeIndex, readDefaults);
```

Comment:
Here is where I pass the existing `typeIndex` through. I realized this when seeing how scalar values are deserialized in other switch statements here.